base/show.jl

```diff
@@ -577,12 +577,16 @@ function show(io::IO, p::Pair)
     iocompact = IOContext(io, :compact => get(io, :compact, true))
     has_tight_type(p) || return show_default(iocompact, p)
 
+    typeinfo = get(io, :typeinfo, Any)
+    typeinfos = typeinfo <: Pair ?
```

Comment:
`&& isa(unwrap_unionall(typeinfo, DataType))`

Reply:
I can only think of `typeinfo = Union{}` where this added test would make a difference, are there other cases? otherwise, we might as well test `typeinfo !== Union{}`?

Reply:
Any Union. And why would we not just test the actual constraint currently imposed by fieldtype?

Reply:
Ah ok thanks, hadn't thought about `Union` of `Pair`s. 

> And why would we not just test the actual constraint currently imposed by fieldtype?

I don't really know what you mean. I should also update this test with `p isa typeinfo` by the way.

Reply:
I'm not familiar enough with `unwrap_unionall` (let alone the constraints imposed by `fieldtype`), but it would seem to me that only testing `isa(unwrap_unionall(typeinfo), DataType)` would not be enough. For all likely cases, testing for `typeinfo !== Union{}` (itself unlikely to happen) would be sufficient for now, and is taken care of by the test `p isa typeinfo` that I just added. I will merge later today if no objection.